dynamic_rest/renderers.py

```diff
@@ -0,0 +1,93 @@
+from collections import defaultdict
+from rest_framework import renderers
+
+
+class DynamicJSONRenderer(renderers.JSONRenderer):
+
+  def render(self, data, accepted_media_type=None, renderer_context=None):
+    # if data is a string, just render it out using the default method
+    # this usually indicates an error message
+    if isinstance(data, basestring):
+      return super(DynamicJSONRenderer, self).render(data, accepted_media_type, renderer_context)
+
+    # populated by _sideload
+    self._data = {}
+
+    # used by _sideload to prevent duplicates
+    self._seen = defaultdict(set)
+
+    # used by _sideload to prevent adding secondary records of the primary resource
+    # from ending up in the primary slot
+    self._plural_name = getattr(
+        renderer_context.get('view').get_serializer().Meta, 'plural_name', 'objects')
+
+    # recursively sideload everything
+    self._sideload(data)
+
+    # add the primary resource data into the response data
+    if isinstance(data, dict):
+      resource_name = getattr(
+          renderer_context.get('view').get_serializer().Meta, 'name', 'object')
+    else:
+      resource_name = self._plural_name
+
+    self._data[resource_name] = data
+
+    view = renderer_context.get('view')
+    if hasattr(view, 'response_meta') and view.response_meta:
+      self._data['meta'] = view.response_meta
+
+    # call super to render the response
+    return super(DynamicJSONRenderer, self).render(self._data, accepted_media_type, renderer_context)
+
+  # recursively traverse the data and extract all resources
+  # that need to be sideloaded
+  def _sideload(self, obj, parent=None, parent_key=None, depth=0):
+    if isinstance(obj, list):
+      # list
+      for key, o in enumerate(obj):
+        # recurse into lists of objects
+        self._sideload(o, parent=obj, parent_key=key, depth=depth)
+    elif isinstance(obj, dict):
+      # object
+
+      if '_model' in obj and '_pk' in obj:
+        model = obj['_model']
+        pk = obj['_pk']
+
+        # strip the model and pk tags from the final response
+        del obj['_pk']
+        del obj['_model']
+
+        # recursively check all fields
+        for key, o in obj.iteritems():
+          if isinstance(o, list) or isinstance(o, dict):
+            # lists or dicts indicate a relation
+            self._sideload(o, parent=obj, parent_key=key, depth=depth + 1)
+
+        seen = True
+        # if this object has not yet been seen
+        if not pk in self._seen[model]:
+          seen = False
+          self._seen[model].add(pk)
+
+        # prevent sideloading the primary objects
+        if depth == 0:
+          return
+
+        # if the primary resource is embedded, add it to a prefixed key
+        if model == self._plural_name:
+          model = '+%s' % model
```

Comment:
Yeah this is an interesting case... I don't really think we should try to optimize for this until we see a user need, because I can see this going one of several ways and it will take some magic to get serializers to be able to reference each other (i.e. there is a circular dependency issue that happens when two serializers attempt to use each other as fields). `cards.alos` makes sense to me. However, it will make things somewhat more difficult for the consumer in the case that the primary resource appears as a secondary in multiple places, e.g. for `GET /alos/?fields[]=cards.alos.&field[]=foo.alos.`, should we return `cards.alos` and also `foo.alos`? The relationships are already in the response, so the consumer should be able to derive them anyway, in which case separate keys might just be a parsing complication.

Let's wait for someone on the frontend to request this and see what they think the behavior should be (my guess is that Alex won't want to support this at all :) ) -- I'll leave a TODO that we should spec the behavior thoroughly.